Fix gene panel importer, add validation and improve documentation

sandertan opened the pull request:
# What? Why?
This PR fixes the gene panel importer (#3191 ), simplifies and clarifies how gene panel study data should be imported and adds validation and tests (#3314).

Changes proposed in this pull request:
- ~Remove option to add `gene_panel:` property to meta file of the mutation file. With this property it was not possible to mark "sequenced but no mutations found"-samples to be on a gene panel. Documentation is updated accordingly.~ I restored this functionality, but updated the documentation to warn users about the limitations.
- Modifications to the importer how gene panel study data is saved to the database
- Add 10 validation rules to check whether gene panel study data and `_sequenced` case lists are correct.
- ~Add 10 unit tests and 1 system test to verify validation rules.~ This is now in a separate PR: https://github.com/thehyve/cbioportal/pull/33 . It will take quite some effort to reduce the number of files changed in this test data PR. It might be better to optimize this at a later time.
- Rewrite documentation for gene panel study data.
- Improve test study file structure

To do:
- [X] Restructure test study to reduce the number of files added in this PR.
- [x] Fix `TestIntegrationTest.java`

# Gene Panel test data
Perhaps also worth mentioning is that I added 15 test samples to study_es_0, to cover all the sequenced / not sequenced / WES / Gene Panel scenarios for mutations / fusions / CNA I could think of. By querying gene `ACPP` with case list `Test profiled samples` you can test this in OncoPrint.

<img width="632" alt="screenshot 2019-01-22 at 23 16 29" src="https://user-images.githubusercontent.com/9624990/51569128-ca0b1980-1e9b-11e9-8355-f4b714061a72.png">

Sjoerd-van-Hagen commented:
@n1zea144 Hi Ben, can you give an indication for when you will have time to review this one? Thanks!

n1zea144 commented on docs/File-Formats.md:
```diff
@@ -1105,58 +1105,44 @@ rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
-3. **gene_list**: Tab separated genes, represented either by all hugo symbols or all entrez_gene_ids.
+To include gene panel data in your instance, the following data and/or configurations can be used:
+1. **Gene panel file**: This file contains the genes on the gene panel. A panel can be used for multiple studies within the instance and should be loaded prior to loading a study with gene panel data. For information on the format and import process please visit: [Import-Gene-Panels](Import-Gene-Panels.md).
+2. **Gene panel matrix file**: This file is used to specify which samples are sequenced on which gene panel in which genetic profile. This required for mutation and fusion data, because the MAF and fusion formats are unable to include samples which are sequenced but no mutations are found. For other genetic data, columns can be added to specify their gene panel, but a property can also be added t ...
```
"This required" should be "This is required".    Suggest "...which are sequenced but no mutations are found" change to "...which are sequenced but contain no called mutations."  Suggest "For other genetic data..." change to "For other genetic profiles..."

sandertan replied:
Done

n1zea144 commented on docs/File-Formats.md:
```diff
@@ -1105,58 +1105,44 @@ rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
+To include gene panel data in your instance, the following data and/or configurations can be used:
+1. **Gene panel file**: This file contains the genes on the gene panel. A panel can be used for multiple studies within the instance and should be loaded prior to loading a study with gene panel data. For information on the format and import process please visit: [Import-Gene-Panels](Import-Gene-Panels.md).
+2. **Gene panel matrix file**: This file is used to specify which samples are sequenced on which gene panel in which genetic profile. This required for mutation and fusion data, because the MAF and fusion formats are unable to include samples which are sequenced but no mutations are found. For other genetic data, columns can be added to specify their gene panel, but a property can also be added t ...
+3. **Gene panel property in meta file**: Adding the `gene_panel:` property to the meta file of a non-mutation data file, will assign all samples from that data type to the gene panel. In this case it is not necessary to include a column for this profile in the gene panel matrix file.
```
Suggest changing "...of a non-mutation data file, will assign all samples from that data type to the gene panel." to "...of a non-mutation data profile, will assign all samples from that profile to the gene panel.".

sandertan replied:
Done

n1zea144 commented on docs/File-Formats.md:
```diff
@@ -1105,58 +1105,44 @@ rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
-
-The second component to gene panel data is associating samples and profile to the panel which applies. The following column is required :
+#### Gene Panel Matrix file
+The gene panel matrix file contains a list of samples in the first column, and an additional column for each profile in the study using the stable_id as the column header. These stable_id's should match the ones in their respective meta files. Fusion events are saved in the mutation table in the cBioPortal database, so they should be included in the `mutations` column. Other than mutation and fus ...
```
Suggest change "Other than mutation and fusion data formats..." to "As described above, genetic profiles other than mutation and fusion data profiles can use..."

sandertan replied:
Done

n1zea144 commented on docs/File-Formats.md:
```diff
@@ -1105,58 +1105,44 @@ rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
+The gene panel matrix file contains a list of samples in the first column, and an additional column for each profile in the study using the stable_id as the column header. These stable_id's should match the ones in their respective meta files. Fusion events are saved in the mutation table in the cBioPortal database, so they should be included in the `mutations` column. Other than mutation and fus ...
 
-- ***SAMPLE_ID***: Sample Id from the study 
+For each sample-profile combination, a gene panel should be specified by using the stable_id from the gene panel file, for example `mutations` for mutation data and `gistic` for discrete CNA data, or `NA` to reflect whole exome. Columns should be separated by tabs.
```
Suggest WES/WGS instead of NA.

sandertan replied:
Done

sheridancbio replied:
What I see currently in the PR is a choice between three options: "WES", "WGS", or "NA". I think what @n1zea144 wanted was to avoid "NA" which would mean something like "we don't know". I also think his request intended that there be a single unique string which represents "either WES or WGS", rather than giving the user an option to distinguish between WES and WGS. So I interpret the request as forbidding "NA", and requiring either a proper gene panel name, or the string "WES/WGS" ... but maybe @n1zea144 can clarify.

n1zea144 replied:
HI @sandertan, sorry for the delay, maybe @sheridancbio let you know I was out of the lab last week.  In any event, I was thinking what Rob said.  Currently, when no gene panel is specified, I believe the frontend (oncoprint) assumes WGS.  Why allow NA when most likely WGS is likely or would result in proper behavior in the frontend.

I was thinking that we distinguish between WGS and WES in the matrix file, but admittedly I'm not sure if we would make much of a distinction between WES or WGS in terms of behavior in the portal.  I do imagine that this information is at hand to a study curator, so why not support  both independently?  A change like this should be coordinated with the frontend.  I would bring @adamabeshouse into the conversation.

adamabeshouse replied:
whatever you choose, we'll be able to handle on the frontend. it just depends on what we want to show, which I would defer to @jjgao for

sandertan replied:
This would require a database change because this info can’t be saved in the gene panel column of the sample_profile table. Perhaps we can move this as an enhancement to a separate github issue?

As mentioned below in separate comment, I restored NA as only possible value. Adding functionality for WES/WGS is I think out of scope for this PR.

I'll squash & rebase if you think it's okay in its current state.

n1zea144 replied:
@sandertan I don't know the use case for introducing NA into the matrix file, but if we are going to allow it, I think we should hear from @jjgao as to what changes need to happen on the frontend so they can be coordinated with this PR.

n1zea144 commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenePanelProfileMap.java:
```diff
@@ -103,69 +101,63 @@ public void run() {
                 }
                 else {
-                    ProgressMonitor.logWarning("No gene panel exists: " + data.get(i));
+                    if (!genePanelName.equals("NA")) {
```
Suggest WES/WGS instead of NA.

sandertan replied:
Done.

sandertan commented on core/src/main/java/org/mskcc/cbio/portal/dao/DaoSampleProfile.java:
```diff
@@ -94,6 +94,37 @@ public static int addSampleProfile(Integer sampleId, Integer geneticProfileId, I
         }
     }
 
+    public static void addGenePanel(Integer sampleId, Integer geneticProfileId, Integer panelId) throws DaoException {
```
I renamed it to `updateSampleProfile`. 

I tested an approach in which the `addSampleProfile` would only do INSERT and the `updateSampleProfile` would only UPDATE, but this lead to an issue for mutation samples that had no mutations found. The `addSampleProfile` function should also be used by the case list importer when the `_sequenced` case list is importer.

I think the best moment to update these functions is when the case list functionality will be refactored. I've added it to the open RFC for that: RFC37: 
Redesigning Case Lists  https://docs.google.com/document/d/1aBbkTAFv5nCqBv66BOgwvlt5kymgt0lQ7l_pqaFrupc/

sheridancbio commented:
@n1zea144 : I've put a card up for reviewing this. We will finish the review process unless you want to look over changes yourself. cc: @cBioPortal/importer-pipelines

jjgao commented:
@sandertan Could you resolve the conflicts? Anything else you need to do before this can be merged?

sheridancbio commented on docs/File-Formats.md:
```diff
@@ -1105,58 +1105,48 @@ rank<TAB>gene<TAB>N<TAB>n<TAB>p<TAB>q
-3. **gene_list**: Tab separated genes, represented either by all hugo symbols or all entrez_gene_ids.
+To include gene panel data in your instance, the following data and/or configurations can be used:
+1. **Gene panel file**: This file contains the genes on the gene panel. A panel can be used for multiple studies within the instance and should be loaded prior to loading a study with gene panel data. For information on the format and import process please visit: [Import-Gene-Panels](Import-Gene-Panels.md).
+2. **Gene panel matrix file**: This file is used to specify which samples are sequenced on which gene panel in which genetic profile. This is required for mutation and fusion data, because the MAF and fusion formats are unable to include samples which are sequenced but contain no called mutations. For other genetic profiles, columns can be added to specify their gene panel, but a property can als ...
```
Ok, I think I see. If you want to link samples to gene panels now, you must do it here explicitly (for mutations and fusions). I guess that was the main purpose of this PR.

sheridancbio commented on core/src/main/java/org/mskcc/cbio/portal/scripts/ImportGenePanelProfileMap.java:
```diff
@@ -39,24 +39,24 @@
     private File genePanelProfileMapFile;
-    private static Properties properties;
     private String cancerStudyStableId;
+    private List<String> NON_GENE_PANEL_VALUES = Arrays.asList("NA", "WES", "WGS");
```
see previous comment ... I think "NA" was requested to be removed as a valid option. My interpretation of the requests from @n1zea144 was that there be only a single value for "non gene panel":
private List<String> NON_GENE_PANEL_VALUES = Collections.singletonList("WES/WGS");
But maybe we should get clarification.

sandertan commented:
Hi @sheridancbio thanks for the review.

1. Regarding your comment on removing functionality for gene panel via property in mutation meta file: I restored that functionality. I didn't know that functionality for the header to contain the samples was in the code. Sorry for missing that, but it wasn't documented/validated. I updated documentation to include it's possible to add `gene_panel` property in meta mutation file, and warned for the limitations by using this. I have not added documentation about the header containing the mutation samples. I discussed this with Pieter, and we think we should have a separate call on this topic. Also, I removed the comments about case lists and sequenced samples. I am actually not sure what the intended functionality is. Added an idea about this later in this comment.

2. Regarding NA / WES / WGS: I kept NA, because else there is no value for samples that are actually NOT sequenced and thus have no gene panel. But indeed it is confusing because we don't have specific functionality to distinguish between WES and WGS. Because this file is about gene panels only, perhaps removing WES/WGS and only allowing NA as non-gene panel value would be best, as this does not suggest that there is any functionality to distinguish between WES and WGS.

Idea for future:
Currently, defining whether a sample is profiled can be defined via case list and saved in the `sample_list` table, but can also be defined via the genetic profile and saved in the `sample_profile` table. This is confusing for developers as well as data curators. Therefore I think we should remove the case lists for genetic profiles, and save and retrieve these using the genetic data files and the `sample_profile` table. Case lists for CNA, expression and proteins are already redundant as they are in the columns of the data. Lists of profiled samples for mutation, fusion and segment data can be added to either a comment line in the data or the meta file. This will greatly simplify the process of preparing staging files, and allows us to remove a lot of complexity (mainly the dependencies between genetic profile data file / case lists / gene panel matrix file).

sandertan commented:
This would require a database change because this info can’t be saved in
the gene panel column of the sample_profile table. Perhaps we can move this
as an enhancement to a separate github issue?

On Mon, 25 Feb 2019 at 17:03, n1zea144 <notifications@github.com> wrote:

> *@n1zea144* commented on this pull request.
> ------------------------------
>
> In docs/File-Formats.md
> <https://github.com/cBioPortal/cbioportal/pull/5602#discussion_r259894301>
> :
>
> >
> -- ***SAMPLE_ID***: Sample Id from the study
> +For each sample-profile combination, a gene panel should be specified by using the stable_id from the gene panel file, for example `mutations` for mutation data and `gistic` for discrete CNA data, or `NA` to reflect whole exome. Columns should be separated by tabs.
>
> HI @sandertan <https://github.com/sandertan>, sorry for the delay, maybe
> @sheridancbio <https://github.com/sheridancbio> let you know I was out of
> the lab last week. In any event, I was thinking what Rob said. Currently,
> when no gene panel is specified, I believe the frontend (oncoprint) assumes
> WGS. Why allow NA when most likely WGS is likely or would result in proper
> behavior in the frontend.
>
> I was thinking that we distinguish between WGS and WES in the matrix file,
> but admittedly I'm not sure if we would make much of a distinction between
> WES or WGS in terms of behavior in the portal. I do imagine that this
> information is at hand to a study curator, so why not support both
> independently? A change like this should be coordinated with the frontend.
> I would bring @adamabeshouse <https://github.com/adamabeshouse> into the
> conversation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cBioPortal/cbioportal/pull/5602#discussion_r259894301>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJLdnm4Xg2XT9LXcP-dp95BlzLtdinc6ks5vRAk3gaJpZM4aAzoB>
> .
>
-- 

Data scientist


thehyve.nl/solutions/cbioportal/

sandertan commented:
@n1zea144 I restored NA as only possible value. Adding additional functionality for WES/WGS is I think out of scope for this PR. 

I'll squash & rebase if you think it's okay in its current state.

Sjoerd-van-Hagen commented:
@n1zea144 hi Ben,

As Sander is in his last days at The Hyve and this PR fixes some issues with the gene panels, reported in the google groups and by our client, I suggest we try to get this merged as soon as possible, at least before he leaves The Hyve. He has a lot of things to wrap up besides this PR.

Any further enhancements can be made in a separate PR.

jjgao commented:
About NA, I agree with @n1zea144 that we should not use "NA" to represent WXS or WGS. I think we should definitely allow WXS and WGS. "NA" might be useful when we really don't know -- but I am wondering if that would ever happen. I think it's probably better to remove NA so that people won't misuse it.

jjgao commented:
@sandertan I guess we are ready to merge this once the tests pass. Could you look into the tests?

Sjoerd-van-Hagen commented:
@jjgao which release will this be in, and when will it be released?

jjgao commented:
@Sjoerd-van-Hagen looks like this was merged into master. We should probably just tab and release as v2.1.1. Let's do it when @inodb is back next week.
